### PR TITLE
PLANNER-865: Workbench: ScoreHolder support in GuidedDecisionTables causes error in server log

### PR DESCRIPTION
**Pull request by mcimbora**
Part of:
- https://github.com/kiegroup/drools/pull/1476
- https://github.com/kiegroup/drools-wb/pull/620
- https://github.com/kiegroup/optaplanner-wb/pull/222
- https://github.com/kiegroup/droolsjbpm-integration/pull/1193

### CODE REVIEW

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatch.java [hunk 1]
```diff
@@ -17,7 +17,8 @@
 package org.optaplanner.workbench.screens.guidedrule.model;
 
 import org.drools.workbench.models.datamodel.rule.IAction;
```
Comment (jomarko) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatch.java]: Unused import

**Review by jomarko — COMMENTED** [review]
@mcimbora it seems that bendable scores causes problem during validation for more details see the comment below. Could you please have a look?

**Comment by jomarko** [comment]
Bendable scores problem:
- Define planning solution like: 
screenshot from 2017-09-19 12-13-32
- Then in some guided decision table define action like: 
screenshot from 2017-09-19 12-16-09
- use this column in table
- Check the source code - generated correctly
- Run validation, there will be validation error

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionBendableHardConstraintMatch.java [hunk 2]
```diff
@@ -62,4 +62,14 @@ public int hashCode() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addHardConstraintMatch(kcontext, ")
+                .append(getPosition())
+                .append(", ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionBendableHardConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionBendableSoftConstraintMatch.java [hunk 3]
```diff
@@ -62,4 +62,14 @@ public int hashCode() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addSoftConstraintMatch(kcontext, ")
+                .append(getPosition())
+                .append(", ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionBendableSoftConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionHardConstraintMatch.java [hunk 4]
```diff
@@ -32,4 +32,12 @@ public ActionHardConstraintMatch(final String constraintMatch) {
     public TemplateAware cloneTemplateAware() {
         return new ActionHardConstraintMatch(getConstraintMatch());
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addHardConstraintMatch(kcontext, ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionHardConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMediumConstraintMatch.java [hunk 5]
```diff
@@ -33,4 +33,12 @@ public ActionMediumConstraintMatch(final String constraintMatch) {
     public TemplateAware cloneTemplateAware() {
         return new ActionMediumConstraintMatch(getConstraintMatch());
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMediumConstraintMatch(kcontext, ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMediumConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableBigDecimalMatch.java [hunk 6]
```diff
@@ -39,4 +39,14 @@ public TemplateAware cloneTemplateAware() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMultiConstraintMatch(kcontext, new java.math.BigDecimal[] {")
+                .append(getActionBendableHardConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("}, new java.math.BigDecimal[] {")
+                .append(getActionBendableSoftConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("})").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableBigDecimalMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableLongMatch.java [hunk 7]
```diff
@@ -39,4 +39,14 @@ public TemplateAware cloneTemplateAware() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMultiConstraintMatch(kcontext, new long[] {")
+                .append(getActionBendableHardConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("}, new long[] {")
+                .append(getActionBendableSoftConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("})").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableLongMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableMatch.java [hunk 8]
```diff
@@ -39,4 +39,14 @@ public TemplateAware cloneTemplateAware() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMultiConstraintMatch(kcontext, new int[] {")
+                .append(getActionBendableHardConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("}, new int[] {")
+                .append(getActionBendableSoftConstraintMatches().stream().map(m -> m.getConstraintMatch()).collect(Collectors.joining(", ")))
+                .append("})").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintBendableMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintHardMediumSoftMatch.java [hunk 9]
```diff
@@ -124,4 +124,16 @@ public int hashCode() {
...
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMultiConstraintMatch(kcontext, ")
+                .append(getActionHardConstraintMatch().getConstraintMatch())
+                .append(", ")
+                .append(getActionMediumConstraintMatch().getConstraintMatch())
+                .append(", ")
+                .append(getActionSoftConstraintMatch().getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintHardMediumSoftMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintHardSoftMatch.java [hunk 10]
```diff
@@ -96,4 +96,14 @@ public int hashCode() {
...
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addMultiConstraintMatch(kcontext, ")
+                .append(getActionHardConstraintMatch().getConstraintMatch())
+                .append(", ")
+                .append(getActionSoftConstraintMatch().getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionMultiConstraintHardSoftMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionSimpleConstraintMatch.java [hunk 11]
```diff
@@ -33,4 +33,12 @@ public ActionSimpleConstraintMatch(final String constraintMatch) {
     public TemplateAware cloneTemplateAware() {
         return new ActionSimpleConstraintMatch(getConstraintMatch());
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addConstraintMatch(kcontext, ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionSimpleConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionSoftConstraintMatch.java [hunk 12]
```diff
@@ -33,4 +33,12 @@ public ActionSoftConstraintMatch(final String constraintMatch) {
     public TemplateAware cloneTemplateAware() {
         return new ActionSoftConstraintMatch(getConstraintMatch());
     }
+
+    @Override
+    public String getStringRepresentation() {
+        return new StringBuilder()
+                .append("scoreHolder.addSoftConstraintMatch(kcontext, ")
+                .append(getConstraintMatch())
+                .append(")").toString();
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/main/java/org/optaplanner/workbench/screens/guidedrule/model/ActionSoftConstraintMatch.java]: Do we not need a ```;``` at the end?

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/test/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatchTest.java [hunk 13]
```diff
@@ -486,4 +486,140 @@ public void cloneIActionMultiConstraintHardSoftMatch() {
...
                      constraintMatch.getActionSoftConstraintMatch().getConstraintMatch());
     }
+
+    @Test
+    public void marshalActionSoftConstraintMatch() {
+        ActionSoftConstraintMatch action = new ActionSoftConstraintMatch("-1");
+
+        String marshaledAction = action.getStringRepresentation();
+
+        assertEquals("scoreHolder.addSoftConstraintMatch(kcontext, -1)",
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/test/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatchTest.java:496]: Formatting.

File: optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/test/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatchTest.java [hunk 14]
```diff
@@ -486,4 +486,140 @@ public void cloneIActionMultiConstraintHardSoftMatch() {
...
+    }
+
+    @Test
+    public void marshalActionBendableSoftConstraintMatch() {
+        ActionBendableSoftConstraintMatch action = new ActionBendableSoftConstraintMatch(1,
+                "-1");
+
+        String marshaledAction = action.getStringRepresentation();
+
+        assertEquals("scoreHolder.addSoftConstraintMatch(kcontext, 1, -1)",
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-api/src/test/java/org/optaplanner/workbench/screens/guidedrule/model/ActionConstraintMatchTest.java]: Formatting.

**Comment by jomarko** [comment]
@manstis ";" is added here https://github.com/kiegroup/drools/pull/1476/commits/86552030d34ec27aeef794dfd2f822de50a53202#diff-4e116fb8c4eec5a7f3fe41645a1b891fR443

**Comment by mcimbora** [comment]
@jomarko When using BendableBigDecimal score, one needs to wrap integer values into java.math.BigDecimal.valueOf(1) (or similar). This will be resolved as a part of https://issues.jboss.org/browse/PLANNER-868.

**Comment by jomarko** [comment]
@mcimbora Thanks Matej. I will approve the ensemble of PRs for PLANER-865.

**Review by manstis — APPROVED** [review]

**Review by jomarko — APPROVED** [review]

**Pull request merged**